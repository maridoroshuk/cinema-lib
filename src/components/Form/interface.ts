import { FormikHelpers } from 'formik';
import { ReactNode } from 'react';

export interface IFormProps<T> {
  children: ReactNode;
  initialValues: T;
  validate: (values: T) => void;
  onFormSubmit: (submitValues: T, { setSubmitting }: FormikHelpers<T>) => void;
}
