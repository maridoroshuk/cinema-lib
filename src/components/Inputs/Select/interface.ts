interface IOptions {
  id: number;
  value: string;
}

export interface ISelectProps {
  config: {
    name: string;
    label: string;
    id: string;
    options: IOptions[];
  };
}
