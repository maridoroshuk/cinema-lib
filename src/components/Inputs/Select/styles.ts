import { Field } from 'formik';
import styled from 'styled-components';

export const Wrapper = styled.div`
  background-color: ${({ theme }) => theme.colors.neutral900};
  padding: ${({ theme }) => `${theme.spaces[30]}px 0`};
  color: ${({ theme }) => theme.colors.neutral000};
  font-family: ${({ theme }) => theme.fontFamily.Poppins};
  font-size: ${({ theme }) => theme.sizes.fontSize[20]}px;
`;

export const Label = styled.label`
  width: 100%;
`;

export const StyledSelect = styled(Field)`
  width: 100%;
  background-color: ${({ theme }) => theme.colors.neutral900};
  border: none;
  outline: none;
  color: ${({ theme }) => theme.colors.neutral000};
  padding: ${({ theme }) => `${theme.spaces[15]}px 0`};
  font-family: ${({ theme }) => theme.fontFamily.Poppins};
  font-size: ${({ theme }) => theme.sizes.fontSize[20]}px;
`;

export const Option = styled.option``;
