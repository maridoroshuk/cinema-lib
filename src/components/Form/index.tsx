import React from 'react';
import { Formik } from 'formik';

import { IFormProps } from './interface';
import { StyledForm } from './styles';

export const Form = <T extends {}>({
  children,
  initialValues,
  validate,
  onFormSubmit,
}: IFormProps<T>) => {
  return (
    <Formik
      initialValues={initialValues}
      validate={validate}
      onSubmit={onFormSubmit}
    >
      <StyledForm>{children}</StyledForm>
    </Formik>
  );
};
