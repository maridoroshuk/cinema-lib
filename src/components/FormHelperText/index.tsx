import React from 'react';

import { IFormHelperTextProps } from './interface';
import { Wrapper } from './styles';

export const FormHelperText = ({ message }: IFormHelperTextProps) => {
  return <Wrapper>{message}</Wrapper>;
};
