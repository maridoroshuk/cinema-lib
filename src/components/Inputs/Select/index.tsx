import React from 'react';

import { ISelectProps } from './interface';
import { Label, Option, StyledSelect, Wrapper } from './styles';

export const Select = ({
  config: { name, label, id, options },
}: ISelectProps) => {
  return (
    <Wrapper>
      <Label htmlFor={id}>
        {label}
        <StyledSelect component="select" name={name} data-testid="select">
          {options.map(({ id, value }) => (
            <Option key={id} value={value}>
              {value}
            </Option>
          ))}
        </StyledSelect>
      </Label>
    </Wrapper>
  );
};
