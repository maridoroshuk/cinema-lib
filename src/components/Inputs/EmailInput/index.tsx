import React, { useId } from 'react';
import { ErrorMessage } from 'formik';

import Send from '@/assets/icons/send.png';
import { Tooltip } from '@/components/Tooltip';

import { Icon, Label, StyledField, SubmitButton, Wrapper } from './styles';
import { IEmailInput } from './interface';

export const EmailInput = ({placeholder}: IEmailInput) => {
  const emailId = useId();

  return (
    <Wrapper>
      <Label htmlFor="email">
        <StyledField
          id={emailId}
          type="email"
          name="email"
          placeholder={placeholder}
          autoComplete="off"
        />
        <ErrorMessage name="email">
          {(message) => <Tooltip>{message}</Tooltip>}
        </ErrorMessage>
      </Label>
      <SubmitButton type="submit">
        <Icon src={Send} alt="send email" />
      </SubmitButton>
    </Wrapper>
  );
};
