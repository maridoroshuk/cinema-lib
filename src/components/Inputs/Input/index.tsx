import React, { useId } from 'react';
import { ErrorMessage } from 'formik';

import { Tooltip } from '@/components/Tooltip';

import { IInputProps } from './interface';
import { AdditionalText, Icon, Label, StyledField, Wrapper } from './styles';

export const Input = ({
  type,
  name,
  icon,
  label,
  placeholder,
  helperText,
  onInputBlur,
}: IInputProps) => {
  const inputId = useId();

  return (
    <Wrapper>
      <Label htmlFor={name}>
        {label}
        {icon && <Icon src={icon} alt={name} />}
        <StyledField
          id={inputId}
          type={type}
          name={name}
          placeholder={placeholder}
          autoComplete="off"
          onBlur={onInputBlur}
        />
        <ErrorMessage name={name}>
          {(message) => <Tooltip>{message}</Tooltip>}
        </ErrorMessage>
        {helperText && <AdditionalText>{helperText}</AdditionalText>}
      </Label>
    </Wrapper>
  );
};
