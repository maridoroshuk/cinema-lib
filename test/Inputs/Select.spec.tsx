import React from 'react';
import { act } from 'react-dom/test-utils';
import { fireEvent, getByTestId, render } from '@testing-library/react';

import { Form } from '@/components/Form';
import { Select } from '@/components/Inputs/Select';

import 'jest-styled-components';

import { setupTestProvider } from '../setupTests';

describe('Select', () => {
  const handleValidate = jest.fn();
  const handleSubmit = jest.fn();

  it('renders with female value', async () => {
    const { container } = render(
      setupTestProvider(
        <Form
          initialValues={{ gender: 'male' }}
          validate={handleValidate}
          onFormSubmit={handleSubmit}
        >
          <Select
            config={{
              name: 'gender',
              label: 'Select gender:',
              id: 'gender-select',
              options: [
                { id: 1, value: 'male' },
                { id: 2, value: 'female' },
                { id: 3, value: 'other' },
              ],
            }}
          />
        </Form>,
      ),
    );

    act(() => {
      fireEvent.change(getByTestId(container, 'select'), {
        target: { value: 'female' },
      });
    });

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveFormValues({ gender: 'female' });
  });

  it('renders', async () => {
    const { container } = render(
      setupTestProvider(
        <Form
          initialValues={{ gender: 'male' }}
          validate={handleValidate}
          onFormSubmit={handleSubmit}
        >
          <Select
            config={{
              name: 'gender',
              label: 'Select gender:',
              id: 'gender-select',
              options: [
                { id: 1, value: 'male' },
                { id: 2, value: 'female' },
                { id: 3, value: 'other' },
              ],
            }}
          />
        </Form>,
      ),
    );

    act(() => {
      fireEvent.change(getByTestId(container, 'select'), {
        target: { value: 'female' },
      });
    });

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveFormValues({ gender: 'female' });
  });
});
