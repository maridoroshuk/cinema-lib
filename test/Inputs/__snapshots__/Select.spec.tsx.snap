// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  background-color: #1E1F27;
  padding: 30px 0;
  color: #FFFFFF;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

.c2 {
  width: 100%;
}

.c3 {
  width: 100%;
  background-color: #1E1F27;
  border: none;
  outline: none;
  color: #FFFFFF;
  padding: 15px 0;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

<form
  action="#"
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="gender-select"
    >
      Select gender:
      <select
        class="c3"
        data-testid="select"
        name="gender"
      >
        <option
          class=""
          value="male"
        >
          male
        </option>
        <option
          class=""
          value="female"
        >
          female
        </option>
        <option
          class=""
          value="other"
        >
          other
        </option>
      </select>
    </label>
  </div>
</form>
`;

exports[`Select renders with female value 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  background-color: #1E1F27;
  padding: 30px 0;
  color: #FFFFFF;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

.c2 {
  width: 100%;
}

.c3 {
  width: 100%;
  background-color: #1E1F27;
  border: none;
  outline: none;
  color: #FFFFFF;
  padding: 15px 0;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

<form
  action="#"
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="gender-select"
    >
      Select gender:
      <select
        class="c3"
        data-testid="select"
        name="gender"
      >
        <option
          class=""
          value="male"
        >
          male
        </option>
        <option
          class=""
          value="female"
        >
          female
        </option>
        <option
          class=""
          value="other"
        >
          other
        </option>
      </select>
    </label>
  </div>
</form>
`;
